File: Ch08.py
# ISLR Ch 8 by Carol Cui

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression, LogisticRegression, LassoCV
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeRegressor, DecisionTreeClassifier, export_graphviz
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor, GradientBoostingClassifier
from sklearn.metrics import confusion_matrix, mean_squared_error

# to use Graphviz
import graphviz
import os
os.environ['PATH'] += os.pathsep + 'C:\\Program Files (x86)\\Graphviz2.38\\bin'

# ----------------------------------------------------------------------------
# Q7
Boston = pd.read_csv('C:\\Users\\Carol\\Desktop\\Boston.csv')
Boston.head()

x = Boston.iloc[:,1:-1]
y = Boston.iloc[:,-1]
xtrain, xtest, ytrain, ytest = train_test_split(x, y, train_size=0.5, random_state=1)

k   = 0
mse = np.zeros((10,6))
for j in range(25, 595, 95):
    for i in range(0,10):
        rf_boston = RandomForestRegressor(n_estimators=j, max_features=i+1, random_state=2)
        rf_boston.fit(xtrain, ytrain)
        ypred     = rf_boston.predict(xtest)
        mse[i,k]  = mean_squared_error(ytest, ypred)
    k += 1

ntree_grid = np.arange(25, 595, 95)
color_opts = ['black','blue','brown','green','red','gray','orange','pink','violet','yellow'] 

fig, ax = plt.subplots(1, 1)
for k in range(0,10):
    ax.plot(ntree_grid, mse[k,:], color=color_opts[k], label='m=%i' %(k+1))
ax.set_xlabel('# of trees')
ax.set_ylabel('test MSE')
ax.set_xticks(ntree_grid)
ax.legend()
# lowest test MSE: feature=7, ntree=25

# ----------------------------------------------------------------------------
# Q8
Carseats = pd.read_csv('C:\\Users\\Carol\\Desktop\\Carseats.csv')
Carseats.head()

Carseats['Urban']     = Carseats.Urban.map({'No':0, 'Yes':1})
Carseats['US']        = Carseats.US.map({'No':0, 'Yes':1})
Carseats['ShelveLoc'] = pd.factorize(Carseats.ShelveLoc)[0]
Carseats              = Carseats.drop('Unnamed: 0', axis=1)

# (a)
x = Carseats.drop(['Sales'], axis=1)
y = Carseats['Sales']
xtrain, xtest, ytrain, ytest = train_test_split(x, y, train_size=0.5, random_state=1)

# (b)
tree_carseats = DecisionTreeRegressor()
tree_carseats.fit(xtrain, ytrain)

ypred = tree_carseats.predict(xtest)
mean_squared_error(ytest, ypred)

dot_data = export_graphviz(tree_carseats, out_file=None, feature_names=xtrain.columns)
graphviz.Source(dot_data)

# (c)
depth = []
for i in range(1,11):
    cv_tree = DecisionTreeRegressor(max_depth=i)
    scores  = cross_val_score(estimator=cv_tree, X=xtrain, y=ytrain, cv=10) # 10-fold CV
    depth.append(scores.mean())
plt.plot(range(1,11), depth)

pruned_tree_carseats = DecisionTreeRegressor(max_depth=depth.index(max(depth))+1)
pruned_tree_carseats.fit(xtrain, ytrain)

ypred = pruned_tree_carseats.predict(xtest)
mean_squared_error(ytest, ypred)

dot_data = export_graphviz(pruned_tree_carseats, out_file=None, feature_names=xtrain.columns)
graphviz.Source(dot_data)

# (d)
bag_carseats = RandomForestRegressor(max_features=len(xtrain.columns), random_state=2)
bag_carseats.fit(xtrain, ytrain)

ypred = bag_carseats.predict(xtest)
mean_squared_error(ytest, ypred)

Importance = pd.DataFrame({'Importance':bag_carseats.feature_importances_*100}, index=xtrain.columns)
Importance.sort_values(by='Importance', axis=0, ascending=True).plot(kind='barh', color='r')
plt.xlabel('Variable Importance')
plt.gca().legend_ = None
# most important: Price, ShelveLoc

# (e)
mse = np.zeros(10)
for i in range(0,10):
    rf_carseats = RandomForestRegressor(max_features=i+1, random_state=2)
    rf_carseats.fit(xtrain, ytrain)
    ypred  = rf_carseats.predict(xtest)
    mse[i] = mean_squared_error(ytest, ypred)
    
plt.plot(range(1,11), mse)
plt.scatter(mse.argmin()+1, min(mse), color='r')
plt.xlabel('# of predictors')   
plt.ylabel('test MSE')
# Optimal # of predictors is 8.

rf_carseats = RandomForestRegressor(max_features=mse.argmin()+1, random_state=2)
rf_carseats.fit(xtrain, ytrain)
Importance = pd.DataFrame({'Importance':rf_carseats.feature_importances_*100}, index=xtrain.columns)
Importance.sort_values(by='Importance', axis=0, ascending=True).plot(kind='barh', color='r')
plt.xlabel('Variable Importance')
plt.gca().legend_ = None
# most important: Price, ShelveLoc

# ----------------------------------------------------------------------------
# Q9
OJ = pd.read_csv('C:\\Users\\Carol\\Desktop\\OJ.csv').drop('Unnamed: 0', axis=1)
OJ.head()

OJ['Purchase'] = pd.factorize(OJ.Purchase)[0]
OJ['Store7']   = OJ.Store7.map({'No':0, 'Yes':1})

# (a)
x = OJ.drop(['Purchase'], axis=1)
y = OJ['Purchase']
xtrain, xtest, ytrain, ytest = train_test_split(x, y, train_size=800, random_state=1)

# (b) & (c)
# use max_depth option to make tree structure more manageable for answering the questions
tree_oj = DecisionTreeClassifier(max_depth=3)
tree_oj.fit(xtrain, ytrain)
print('training error rate: %.2f' %(1-tree_oj.score(xtrain, ytrain))) 
# The tree has 15% training error rate.

# parse the tree structure
n_nodes        = tree_oj.tree_.node_count
children_left  = tree_oj.tree_.children_left
children_right = tree_oj.tree_.children_right
feature        = tree_oj.tree_.feature
threshold      = tree_oj.tree_.threshold

node_depth = np.zeros(shape=n_nodes, dtype=np.int64)
is_leaves  = np.zeros(shape=n_nodes, dtype=bool)
stack = [(0, -1)]  # seed is the root node id and its parent depth
while len(stack) > 0:
    node_id, parent_depth = stack.pop()
    node_depth[node_id]   = parent_depth + 1

    # if we have a test node
    if (children_left[node_id] != children_right[node_id]):
        stack.append((children_left[node_id], parent_depth + 1))
        stack.append((children_right[node_id], parent_depth + 1))
    else:
        is_leaves[node_id] = True

print("The binary tree structure has %s nodes and has the following tree structure:" %n_nodes)
for i in range(n_nodes):
    if is_leaves[i]:
        print("%snode=%i leaf node." %(node_depth[i] * "\t", i))
    else:
        print("%snode=%i test node: go to node %i if X[:, %i] <= %.2f else to node %i."
              %(node_depth[i] * "\t", i, children_left[i], feature[i], threshold[i], children_right[i],))

print('# of terminal nodes: %i' %sum(is_leaves))
# The tree has 8 terminal nodes.

# (d)
dot_data = export_graphviz(tree_oj, out_file=None, feature_names=xtrain.columns)
graphviz.Source(dot_data)

# (e)
ypred = tree_oj.predict(xtest)
cmat  = pd.DataFrame(confusion_matrix(ytest, ypred).T, index = ['No', 'Yes'], columns = ['No', 'Yes'])
print(cmat)
# test error rate = (31+18)/270 = 18%

# (f)
depth = np.zeros(10)
for i in range(1,11):
    cv_tree    = DecisionTreeClassifier(max_depth=i)
    scores     = cross_val_score(estimator=cv_tree, X=xtrain, y=ytrain, cv=10) # 10-fold CV
    depth[i-1] = scores.mean()
print('optimal depth: %i' %(depth.argmax()+1)) #4

# (g)
plt.plot(range(1,11), 1-depth)
plt.xlabel('tree depth')
plt.ylabel('cv error rate')

# (h) best depth: 4

# (i)
pruned_tree_oj = DecisionTreeClassifier(max_depth=depth.argmax()+1)
pruned_tree_oj.fit(xtrain, ytrain)

dot_data = export_graphviz(pruned_tree_oj, out_file=None, feature_names=xtrain.columns)
graphviz.Source(dot_data)

# (j)
unpruned_tree_oj = DecisionTreeClassifier()
unpruned_tree_oj.fit(xtrain, ytrain)

print('unpruned training error rate: %.2f' %(1-unpruned_tree_oj.score(xtrain, ytrain))) 
print('pruned training error rate: %.2f' %(1-pruned_tree_oj.score(xtrain, ytrain))) 
# Pruned tree has 15% training error rate, much higher than unpruned tree.

# (k)
ypred = unpruned_tree_oj.predict(xtest)
cmat1 = pd.DataFrame(confusion_matrix(ytest, ypred).T, index = ['No', 'Yes'], columns = ['No', 'Yes'])
print(cmat1)
# unpruned tree test error rate = (41+24)/270 = 24%

ypred = pruned_tree_oj.predict(xtest)
cmat2 = pd.DataFrame(confusion_matrix(ytest, ypred).T, index = ['No', 'Yes'], columns = ['No', 'Yes'])
print(cmat2)
# pruned tree test error rate = (33+18)/270 = 19%, lower than unpruned tree

# ----------------------------------------------------------------------------
# Q10
# (a)
Hitters = pd.read_csv('C:\\Users\\Carol\\Desktop\\Hitters.csv').drop('Unnamed: 0', axis=1)
Hitters = Hitters.dropna()
Hitters.head()

Hitters['League']    = Hitters.League.map({'A':0, 'N':1})
Hitters['NewLeague'] = Hitters.NewLeague.map({'A':0, 'N':1})
Hitters['Division']  = Hitters.Division.map({'E':0, 'W':1})
Hitters['logSal']    = np.log(Hitters.Salary)

# (b)
x = Hitters.drop(['Salary','logSal'], axis=1)
y = Hitters['logSal']
xtrain, xtest, ytrain, ytest = train_test_split(x, y, train_size=200, random_state=1)

# (c)
lambda_grid = np.linspace(0.001,0.201,20)
mse         = np.zeros(20)
for i in range(0,20):
    boost_hitters = GradientBoostingRegressor(n_estimators=1000, learning_rate=lambda_grid[i], random_state=2)
    boost_hitters.fit(xtrain, ytrain)
    mse[i] = mean_squared_error(ytrain, boost_hitters.predict(xtrain))

plt.plot(lambda_grid, mse)
plt.xlabel('lambda')
plt.ylabel('training MSE') 
plt.xticks(lambda_grid)

# (d)
lambda_grid = np.linspace(0.001,0.201,20)
mse         = np.zeros(20)
for i in range(0,20):
    boost_hitters = GradientBoostingRegressor(n_estimators=1000, learning_rate=lambda_grid[i], random_state=2)
    boost_hitters.fit(xtrain, ytrain)
    mse[i] = mean_squared_error(ytest, boost_hitters.predict(xtest))

plt.plot(lambda_grid, mse)
plt.xlabel('lambda')
plt.ylabel('test MSE') 
plt.xticks(lambda_grid)

# (e)
# multiple linear regression from Ch3
lm_hitter = LinearRegression().fit(xtrain, ytrain)
lm_mse    = mean_squared_error(ytest, lm_hitter.predict(xtest)) 

# lasso from Ch6
lasso_hitter = LassoCV(cv=10, normalize=True).fit(xtrain, ytrain)
lasso_mse    = mean_squared_error(ytest, lasso_hitter.predict(xtest)) 

# plot
x_label = np.arange(3)
plt.bar(x_label, [max(mse), lm_mse, lasso_mse])
plt.xticks(x_label, ('max boost', 'lm', 'lasso'))
plt.ylabel('test MSE')
# Boosting performs the best.

# (f)
best_lambda = lambda_grid[mse.argmin()]
best_boost_hitters = GradientBoostingRegressor(n_estimators=1000, learning_rate=best_lambda, random_state=2)
best_boost_hitters.fit(xtrain, ytrain)

f_imp   = best_boost_hitters.feature_importances_*100
rel_imp = pd.Series(f_imp, index=xtrain.columns).sort_values(inplace=False)
rel_imp.T.plot(kind='barh', color='r')
plt.xlabel('Variable Importance')
plt.gca().legend_ = None
# CAtBat is the most important predictor. 

# (g)
bag_hitters = RandomForestRegressor(max_features=len(xtrain.columns), random_state=2)
bag_hitters.fit(xtrain, ytrain)
bag_mse     = mean_squared_error(ytest, bag_hitters.predict(xtest))

# ----------------------------------------------------------------------------
# Q11
Caravan = pd.read_csv('C:\\Users\\Carol\\Desktop\\Caravan.csv').drop('Unnamed: 0', axis=1)
Caravan.head()

Caravan['Purchase'] = Caravan.Purchase.map({'No':0, 'Yes':1})

# (a)
x = Caravan.drop(['Purchase'], axis=1)
y = Caravan['Purchase']
xtrain, xtest, ytrain, ytest = train_test_split(x, y, train_size=1000, random_state=1)

# (b)
boost_caravan = GradientBoostingClassifier(n_estimators=1000, learning_rate=0.01, random_state=2)
boost_caravan.fit(xtrain, ytrain)

f_imp   = boost_caravan.feature_importances_*100
rel_imp = pd.Series(f_imp, index=xtrain.columns).sort_values(inplace=False)
rel_imp.T.plot(kind='barh', color='r')
plt.xlabel('Variable Importance')
plt.gca().legend_ = None
# MGODOV is the most important predictor.

# (c)
# boost
ypred_prob = boost_caravan.predict_proba(xtest)[:,1]
ypred      = np.where(ypred_prob>0.2, 1, 0)

cmat1 = pd.DataFrame(confusion_matrix(ytest, ypred).T, index = ['No', 'Yes'], columns = ['No', 'Yes'])
print(cmat1)
# % of ppl predicted to purchase actually purchase: 48/290 = 16.55%

# logit
lm_caravan = LogisticRegression().fit(xtrain, ytrain)
ypred_prob = lm_caravan.predict_proba(xtest)[:,1]
ypred      = np.where(ypred_prob>0.2, 1, 0)

cmat2 = pd.DataFrame(confusion_matrix(ytest, ypred).T, index = ['No', 'Yes'], columns = ['No', 'Yes'])
print(cmat2)
# % of ppl predicted to purchase actually purchase: 60/350 = 17.14%

# KNN
knn_caravan = KNeighborsClassifier(n_neighbors=5).fit(xtrain, ytrain)
ypred_prob  = knn_caravan.predict_proba(xtest)[:,1]
ypred       = np.where(ypred_prob>0.2, 1, 0)

cmat3 = pd.DataFrame(confusion_matrix(ytest, ypred).T, index = ['No', 'Yes'], columns = ['No', 'Yes'])
print(cmat3)
# % of ppl predicted to purchase actually purchase: 36/183 = 19.67%

# Boosting performs the best, while KNN performs the worst.

# ----------------------------------------------------------------------------
# Q12
College = pd.read_csv('C:\\Users\\Carol\\Desktop\\College.csv').drop('Unnamed: 0', axis=1)
College['Private'] = College.Private.map({'No':0, 'Yes':1})

x = College.drop(['Grad.Rate'], axis=1)
y = College['Grad.Rate']
xtrain, xtest, ytrain, ytest = train_test_split(x, y, train_size=0.5, random_state=1)

# boost
boost_college = GradientBoostingRegressor(random_state=2)
boost_college.fit(xtrain, ytrain)
boost_mse = mean_squared_error(ytest, boost_college.predict(xtest))

# bag
bag_college = RandomForestRegressor(max_features=len(xtrain.columns), random_state=3)
bag_college.fit(xtrain, ytrain)
bag_mse = mean_squared_error(ytest, bag_college.predict(xtest)) 

# random forest
# Default max_features is p/3 in R, so use 6 here.
rf_college = RandomForestRegressor(max_features=6, random_state=4)
rf_college.fit(xtrain, ytrain)
rf_mse = mean_squared_error(ytest, rf_college.predict(xtest)) 

# linear
lm_college = LinearRegression().fit(xtrain, ytrain)
lm_mse     = mean_squared_error(ytest, lm_college.predict(xtest)) 

# plot
x_label = np.arange(4)
plt.bar(x_label, [boost_mse, bag_mse, rf_mse, lm_mse])
plt.xticks(x_label, ('boost', 'bag', 'rf', 'lm'))
plt.ylabel('test MSE')

# Linear regression performs the best.

File: Ch08.R
# Ch 8 Applied
# -------------------------------------------
# Q7
rm(list=ls())

library(MASS)
fix(Boston)

set.seed(1) 
train       = sample(1:nrow(Boston), nrow(Boston)/2)
boston.test = Boston[-train,'medv']

library(randomForest)

k     = 0
error = matrix(0, 10, 6)
for (j in seq(25,500,95)){
  k = k+1
  for (i in 1:10){
    rf.boston  = randomForest(medv~., data=Boston, subset=train, mtry=i, ntree=j)
    yhat.rf    = predict(rf.boston, newdata=Boston[-train,]) 
    error[i,k] = mean((yhat.rf-boston.test)^2)
  }
}

pred.num   = seq(1, 10)
ntree.grid = seq(25, 500, 95)
color.opts = c('black','blue','brown','green','red','gray','orange','pink','violet','yellow') 

dev.new()
plot(ntree.grid, error[1,], xaxt='n', xlab='# of trees', ylab='test MSE', type='l', ylim=c(floor(min(error)),ceiling(max(error)))) 
for (i in 2:10){lines(ntree.grid, error[i,], col=color.opts[i])}
axis(1, at=ntree.grid, las=2)
legend('topright', legend=paste0('m = ',pred.num), col=color.opts, lty=1)
# lowest test MSE: mtry=5, ntree=215

# -------------------------------------------
# Q8
rm(list=ls())

library(ISLR)
fix(Carseats)

# (a)
set.seed(1) 
train    = sample(1:nrow(Carseats), nrow(Carseats)/2)
car.test = Carseats[-train,'Sales']

# (b)
library(tree)
tree.car = tree(Sales~., Carseats, subset=train)
summary(tree.car)

dev.new()
plot(tree.car) 
text(tree.car, pretty=0)

# test MSE
tree.pred = predict(tree.car, newdata=Carseats[-train,]) 
mean((tree.pred-car.test)^2) #4.92

# (c)
set.seed(2)
cv.car     = cv.tree(tree.car) 
cv.size    = cv.car$size[which.min(cv.car$dev)]
prune.car  = prune.tree(tree.car, best=cv.size) 
prune.pred = predict(prune.car, newdata=Carseats[-train,]) 
mean((prune.pred-car.test)^2) #4.92
# Pruning doesn't improve test MSE cuz it selects the most complex tree. 

# (d)
set.seed(3) 
bag.car  = randomForest(Sales~., data=Carseats, subset=train, mtry=10, importance=T) 
bag.pred = predict(bag.car, newdata=Carseats[-train,]) 
mean((bag.pred-car.test)^2) #2.63

importance(bag.car)
varImpPlot(bag.car)
# most important: Price, ShelveLoc

# (e)
set.seed(3) 
rf.mse = rep(0,10)
for (i in 1:10){
  rf.car    = randomForest(Sales~., data=Carseats, subset=train, mtry=i, importance=T)
  rf.pred   = predict(rf.car, newdata=Carseats[-train,]) 
  rf.mse[i] = mean((rf.pred-car.test)^2)
}
 
dev.new()
plot(rf.mse, xlab='# of predictors', ylab='test MSE', type='l') 
points(which.min(rf.mse), min(rf.mse), col='red', cex=2, pch=20)
# Optimal # of predictors is 8.

rf.car = randomForest(Sales~., data=Carseats, subset=train, mtry=which.min(rf.mse), importance=T)
importance(rf.car)
varImpPlot(rf.car)
# most important: Price, ShelveLoc

# -------------------------------------------
# Q9
rm(list=ls())

library(ISLR)
fix(OJ)

# (a)
set.seed(1) 
train   = sample(1:nrow(OJ), 800)
oj.test = OJ[-train,'Purchase']

# (b)
tree.oj = tree(Purchase~., OJ, subset=train)
summary(tree.oj)
# The tree has 15.88% training error rate and 9 terminal nodes.

# (c)
tree.oj
# last node: split criterion = LoyalCH>0.764572, 261 obs in the branch, dev=91.2,
#            overall prediction = CH, (95.79% Yes, 4.22% No)

# (d)
dev.new()
plot(tree.oj) 
text(tree.oj, pretty=0)

# (e)
tree.pred = predict(tree.oj, OJ[-train,], type='class')
table(tree.pred, oj.test)
# test error rate = (38+8)/270 = 17%

# (f)
set.seed(2) 
cv.oj = cv.tree(tree.oj, FUN=prune.misclass)
cv.oj

# (g)
plot(cv.oj$size, cv.oj$dev, type='b')

# (h) best size: 8 or 9

# (i)
prune.oj = prune.misclass(tree.oj, best=5) 
plot(prune.oj) 
text(prune.oj, pretty=0)

# (j)
summary(prune.oj)
# Pruned tree has 16.25% training error rate, higher than unpruned tree.

# (k)
prune.pred = predict(prune.oj, OJ[-train,], type='class')
table(prune.pred, oj.test)
# test error rate = (36+8)/270 = 16%, lower than unpruned tree

# -------------------------------------------
# Q10
rm(list=ls())

library(ISLR)
fix(Hitters)

# (a)
Hitters        = na.omit(Hitters)
Hitters$logSal = log(Hitters$Salary)

# (b)
set.seed(1) 
train    = sample(1:nrow(Hitters), 200)
hit.test = Hitters[-train,'logSal']

# (c)
library(gbm)

set.seed(1) 
j         = 0
train.mse = rep(0, 20)
for (i in seq(0.001,0.201,0.01)){
  j = j + 1
  boost.hit    = gbm(logSal~.-Salary, data=Hitters[train,], distribution='gaussian', n.trees=1000, shrinkage=i)
  boost.pred   = predict(boost.hit, newdata=Hitters[train,], n.trees=1000, shrinkage=i) 
  train.mse[j] = mean((boost.pred-Hitters[train,'logSal'])^2)
}

lambda.grid = seq(0.001,0.201,0.01)
plot(lambda.grid, train.mse, xaxt='n', xlab='lambda', ylab='train MSE', type='b') 
axis(1, at=lambda.grid, las=2)

# (d)
set.seed(1) 
j        = 0
test.mse = rep(0, 20)
for (i in seq(0.001,0.201,0.01)){
  j = j + 1
  boost.hit   = gbm(logSal~.-Salary, data=Hitters[train,], distribution='gaussian', n.trees=1000, shrinkage=i)
  boost.pred  = predict(boost.hit, newdata=Hitters[-train,], n.trees=1000, shrinkage=i) 
  test.mse[j] = mean((boost.pred-hit.test)^2)
}

lambda.grid = seq(0.001,0.201,0.01)
plot(lambda.grid, test.mse, xaxt='n', xlab='lambda', ylab='test MSE', type='b') 
axis(1, at=lambda.grid, las=2) # best 0.011

# (e)
# multiple linear regression from Ch3
lm.hit  = lm(logSal~.-Salary, data=Hitters[train,])
lm.pred = predict(lm.hit, Hitters[-train,])
lm.mse  = mean((lm.pred-hit.test)^2) #0.48

# lasso from Ch6
library(glmnet)
x.train = model.matrix(logSal~.-Salary, Hitters[train,])[,-1]
y.train = Hitters[train,'logSal']
x.test  = model.matrix(logSal~.-Salary, Hitters[-train,])[,-1] 

set.seed(1) 
cv.out     = cv.glmnet(x.train, y.train, alpha=1)
bestlam    = cv.out$lambda.min
lasso.pred = predict(cv.out, s=bestlam, newx=x.test) 
lasso.mse  = mean((lasso.pred-hit.test)^2) #0.47

barplot(c(max(test.mse), lm.mse, lasso.mse), xaxt='n', ylab='test MSE')
axis(1, at=1:3, labels=c('max boosting', 'lm', 'lasso'))
# Boosting performs the best.

# (f)
best.boost.hit = gbm(logSal~.-Salary, data=Hitters[train,], distribution='gaussian', n.trees=1000, shrinkage=0.011)
summary(best.boost.hit)
# CRuns is the most important predictor. 

# (g)
library(randomForest) 
set.seed(1) 
bag.hit  = randomForest(logSal~.-Salary, data=Hitters[train,], mtry=19) 
bag.pred = predict(bag.hit, newdata=Hitters[-train,]) 
bag.mse  = mean((bag.pred-hit.test)^2) #0.25

# -------------------------------------------
# Q11
rm(list=ls())

library(ISLR)
fix(Caravan)

# (a)
Caravan$Pur01 = ifelse(Caravan$Purchase=='Yes', 1, 0)

set.seed(1) 
train    = sample(1:nrow(Caravan), 1000)
can.test = Caravan[-train,'Pur01']

# (b)
boost.can = gbm(Pur01~.-Purchase, data=Caravan[train,], distribution='bernoulli', n.trees=1000, shrinkage=0.01)
summary(boost.can)
# PPERSAUT is the most important predictor. 

# (c)
boost.pred   = predict(boost.can, newdata=Caravan[-train,], distribution='bernoulli', n.trees=1000, shrinkage=0.01) 
boost.pred01 = rep(0, nrow(Caravan[-train,])) 
boost.pred01[boost.pred>.2] = 1
table(boost.pred01,can.test)
# % of ppl predicted to purchase actually purchase: 0%

# logit
glm.can  = glm(Pur01~.-Purchase, data=Caravan[train,], family=binomial)
glm.prob = predict(glm.can, Caravan[-train,], type='response')
glm.pred = rep(0, nrow(Caravan[-train,]))  
glm.pred[glm.prob>.2] = 1
table(glm.pred,can.test)
# % of ppl predicted to purchase actually purchase: 52/(319+52) = 14.02%

# KNN
library(class) 
x.train = model.matrix(Pur01~.-Purchase, Caravan[train,])[,-1]
y.train = Caravan[train,'Pur01']
x.test  = model.matrix(Pur01~.-Purchase, Caravan[-train,])[,-1]

set.seed(1) 
knn.pred = knn(x.train, x.test, y.train, k=5, prob=T)
knn.prob = attr(knn.pred, 'prob')
new.pred = knn.pred
# use 0.2 as the probability threshold for Purchase
for (i in 1:nrow(x.test)){
  if (knn.pred[i] == 0 & (1-knn.prob[i])>0.2){
    new.pred[i] = 1
  }
}
table(new.pred,can.test)
# % of ppl predicted to purchase actually purchase: 27/(137+27) = 16.46%
# Logit performs the best, while boosting performs the worst. 

# -------------------------------------------
# Q12
rm(list=ls())

library(ISLR)
fix(College)

set.seed(1)
train = sample(1:nrow(College), nrow(College)/2)

# boosting
library(gbm)
boost.fit  = gbm(Grad.Rate~., data=College[train,], distribution='gaussian', n.trees=1000)
boost.pred = predict(boost.fit, newdata=College[-train,], n.trees=1000) 
boost.mse  = mean((boost.pred-College[-train,'Grad.Rate'])^2) #194.58

# bagging
library(randomForest)
set.seed(2) 
bag.fit  = randomForest(Grad.Rate~., data=College[train,], mtry=17) 
bag.pred = predict(bag.fit, newdata=College[-train,]) 
bag.mse  = mean((bag.pred-College[-train,'Grad.Rate'])^2) #164.43

# random forest
set.seed(3) 
rf.fit  = randomForest(Grad.Rate~., data=College[train,])
rf.pred = predict(rf.fit, newdata=College[-train,]) 
rf.mse  = mean((rf.pred-College[-train,'Grad.Rate'])^2) #160.57

# linear
lm.fit  = lm(Grad.Rate~., data=College[train,])
lm.pred = predict(lm.fit, College[-train,])
lm.mse  = mean((lm.pred-College[-train,'Grad.Rate'])^2) #168.47

# plot
barplot(c(boost.mse, bag.mse, rf.mse, lm.mse), xaxt='n', ylab='test MSE')
axis(1, at=1:4, labels=c('boost', 'bag', 'rf', 'lm'))
# Random forest w/ default parameters performs the best.
